Date: Thu, 7 Jul 2016 10:09:40 +0600
Subject: [PATCH 3/5] Better performance for CharBuffer with String as input

String is native in ART so CharAt is very expensive if you access it
throudh CharSequence interface. CharBuffer.wrap creates a CharBuffer where
access to each element of String occurs through CharAt method.
The patch forces conversion of String to char[] to speedup input access
if the underlying source for CharBuffer is String.
We will need to allocate a new array but access is significantly improved.
This speedup of java.nio realibility test suite.

---
 ojluni/src/main/java/java/nio/CharBuffer.java | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/ojluni/src/main/java/java/nio/CharBuffer.java b/ojluni/src/main/java/java/nio/CharBuffer.java
index 8a5c64f..09de39e 100644
--- a/ojluni/src/main/java/java/nio/CharBuffer.java
+++ b/ojluni/src/main/java/java/nio/CharBuffer.java
@@ -262,6 +262,11 @@ public abstract class CharBuffer
      */
     public static CharBuffer wrap(CharSequence csq, int start, int end) {
         try {
+            if (csq instanceof String) {
+              // String is native, so access to element by CharAt will take a long
+              // time, so use it as char[].
+              return new HeapCharBuffer(((String)csq).toCharArray(), start, end - start, true);
+            }
             return new StringCharBuffer(csq, start, end);
         } catch (IllegalArgumentException x) {
             throw new IndexOutOfBoundsException();
-- 
1.9.1

